### PR TITLE
Some aggregator improvements and better stats

### LINKED ISSUE #356: when aggregations run behind, the random limiter acquisitions cause data to be dropped
#354 introduced limiting of concurrent aggregations.
if a backlog accumulates of aggregations waiting for a slot, that backlog is processed in random order, so newer aggregations for a given key may be processed before an older one. this results in out of order data, and data being dropped.

we should:
1) process the queue in order
2) see why they're processing so slowly. in my tests i was using simple count aggregation which should all return very promptly. and flushing into table.In (which aggregates to the routes) should also be promptly.

### PR DESCRIPTION
**Pull request by Dieterbe**
fix #356 
because of that, may also help with #241

### CODE REVIEW

**Comment by replay** [comment]
Instead of copying the `stats/*` files from MT, why not just vendor MT and use it like this?

**Comment by Dieterbe** [comment]
1) sometimes a little copying is better than a little dependency
2) metrictank is actually a huge dependency. note that dep can't vendor subpackages, it would want to vendor all of metrictank, which also comes with a large amount of dependencies

File: aggregator/aggregator.go [hunk 1]
```diff
@@ -157,24 +197,48 @@ func (a *Aggregator) AddOrCreate(key string, ts uint32, quantized uint, value fl
...
+		pos = i
+	}
+	// now we must delete all the timestamps from the ordered list
+	if pos == -1 {
+		// we didn't process anything, so no action needed
+		return
 	}
+	if pos == len(a.tsList)-1 {
+		// we went through all of them. can just reset the slice
+		a.tsList = a.tsList[:1]
```
Comment (replay) [inline aggregator/aggregator.go]: is this supposed to completely reset the slice? shouldn't it be `a.tsList = a.tsList[:0]`?

File: aggregator/aggregator.go [hunk 2]
```diff
@@ -120,26 +129,57 @@ func NewMocked(fun, regex, prefix, sub, outFmt string, cache bool, interval, wai
...
+			// if both levels exist, we can just add the value and that's it
+			proc.Add(value, ts)
+		}
 	} else {
+		// first level doesn't exist. create it and add the ts to the list
+		// (second level will be created below)
+		a.tsList = append(a.tsList, quantized)
+		if len(a.tsList) > 1 && a.tsList[len(a.tsList)-2] > quantized {
+			sort.Sort(TsSlice(a.tsList))
+		}
```
Comment (DanCech) [inline aggregator/aggregator.go]: Since we keep `a.tsList` sorted, it may be more efficient to use https://github.com/golang/go/wiki/SliceTricks#insert than to re-sort it, though I suppose in the vast majority of cases the newly-appended value would already be the greatest.

File: aggregator/aggregator.go [hunk 3]
```diff
@@ -157,24 +197,48 @@ func (a *Aggregator) AddOrCreate(key string, ts uint32, quantized uint, value fl
...
 	}
+	if pos == len(a.tsList)-1 {
+		// we went through all of them. can just reset the slice
+		a.tsList = a.tsList[:1]
+		return
+	}
+
+	copy(a.tsList[0:], a.tsList[pos+1:])
+	a.tsList = a.tsList[:len(a.tsList)-pos-1]
+
```
Comment (DanCech) [inline aggregator/aggregator.go]: Looking at this whole block, would it make more sense to do something like:
```
for quantized, ok := a.tsList[0]; ok && quantized < ts {
  // process a.aggregations[quantized]
  delete(a.aggregations, quantized)
  a.tsList = a.tsList[1:]
}
```
Also not sure why the current code uses `a.aggregations[ts]` vs `a.aggregations[quantized]`

File: aggregator/aggregator.go [hunk 4]
```diff
@@ -157,24 +197,48 @@ func (a *Aggregator) AddOrCreate(key string, ts uint32, quantized uint, value fl
...
+	flushWaiting.Inc(1)
 	flushes.Add()
-	for k, proc := range a.aggregations {
-		if k.ts <= ts {
+	flushWaiting.Dec(1)
+	defer flushes.Done()
+
+	pos := -1 // will track the pos of the last ts position that was successfully processed
+	for i, quantized := range a.tsList {
+		if quantized > ts {
```
Comment (replay) [inline aggregator/aggregator.go]: if it's `>` instead of `>=` this makes the `ts` inclusive at the end. but when the `ts` is calculated on line `:343` (`now.Add(-time.Duration(a.Wait) * time.Second)`) it only subtracts the wait time from the current second, so shouldn't that condition be `>=`? Otherwise if we just passed the second border then this condition implements the wait time to only be at least `ts-1`.

File: aggregator/aggregator.go (hunk 1 above)
Comment (Dieterbe) [reply aggregator/aggregator.go]: oops

File: aggregator/aggregator.go (hunk 2 above)
Comment (Dieterbe) [reply aggregator/aggregator.go]: re may be. yes, maybe..
re vast majority value already greatest, yes
doesn't seem worth digging into this deeper.

File: aggregator/aggregator.go (hunk 3 above)
Comment (Dieterbe) [reply aggregator/aggregator.go]: 1) what replay said. don't want to waste slots in the beginning of the slice. will document better
2) `a.aggregations[ts]` this is a bug! will fix . actually will rename ts to cutoff and quantized to ts.
3) re moving the timestamp comparison to the for condition. i think both styles have their proponents, unless someone can link to an established style guide that we can implement, let's focus on things that matter more.

Comment (DanCech) [reply aggregator/aggregator.go]: 1. my understanding is that go actually handles that by copying when required to keep the array reasonable, at least per https://grokbase.com/t/gg/golang-nuts/136crhcaw1/go-nuts-slice-remove-element-memory-question (it's also the approach outlined in https://github.com/golang/go/wiki/SliceTricks#pop-frontshift
2. cool
3. Right now the logic for how we update `a.tsList` is convoluted, and the cleanup is separated from the `delete` from `a.aggregations`.  The intent is to remove that complexity and make the code easier to follow, and remove the processed timestamp from both `a.tsList` and `a.aggregations` at the same time.

File: aggregator/aggregator.go (hunk 4 above)
Comment (Dieterbe) [reply aggregator/aggregator.go]: This description a bit hard to follow, but let's turn this into an example which i think demonstrates the problem you're describing, but in a clearer way.
"if we just passed the second border", so for example now = 15:00:42,001 (3 PM, 42 sec and 1 ms)
wait = 22s
this makes thresh = 15:00:20,001, after `.Unix()`, Flush is called with a timestamp of 15:00:20). 
imagine we have a quantized value of 15:00:20, `quantized > ts` is false so we do not break and process/flush this aggregate.
If i'm reading your argument right, you're saying this value should not be flushed because one more second should have passed for it to be flushed.
But from what I can see, 22 seconds and 1 ms have passed, so the wait period has passed so we should flush.

Comment (Dieterbe) [reply aggregator/aggregator.go]: Also FYI this was very recently changed. see #279 for more information

File: aggregator/aggregator.go (hunk 3 above)
Comment (Dieterbe) [reply aggregator/aggregator.go]: 3/1 fair enough, we can look for ways to simplify.. but `a.tsList = a.tsList[1:]` doesn't seem like a good choice because it'll allocate a new backing array and copy each time we process a `ts`. this generates lots of garbage and overhead (btw wow when did the compiler start doing this). The "example 2" of the first link is basically a more elegant (though also more opaque) way of writing copy statements.  But.. do we want to trade a single copy for O(num-ts-being-processed) copies?  Luckily normally num-ts-being-processed should be very low (typically 1), so that makes it a reasonable decision. but i would rather write them as explicit copies rather than relying than append implementation details.

Comment (DanCech) [reply aggregator/aggregator.go]: The new backing array isn't allocated when you slice, but rather when you `append`, in the current code you're still going to get the allocation when we append regardless of how we handle things here, so if I'm understanding things correctly it'll be more efficient to use `a.tsList = a.tsList[1:]` here and let golang handle re-allocating the backing array when we append on line 176.  Am I missing something?

Comment (Dieterbe) [reply aggregator/aggregator.go]: Oh and btw `quantized, ok := a.tsList[0]` doesn't work on slices.

Comment (DanCech) [reply aggregator/aggregator.go]: Lots of ways to skin that cat, eg:
```
for len(a.tsList) > 0 && a.tsList[0] < cutoff {
  quantized := a.tsList[0]
```

Comment (robert-milan) [reply aggregator/aggregator.go]: It won't always grow when we append, it depends on how much capacity we have left.
To demonstrate: https://play.golang.org/p/cIOo33SEqff

File: clock/clock.go [hunk 5]
```diff
@@ -13,11 +13,11 @@ import (
...
 // to scheduling etc.
-func AlignedTick(period, offset time.Duration) <-chan time.Time {
+func AlignedTick(period, offset time.Duration, bufSize int) <-chan time.Time {
 	// note that time.Ticker is not an interface,
 	// and that if we instantiate one, we can't write to its channel
 	// hence we can't leverage that type.
-	c := make(chan time.Time)
+	c := make(chan time.Time, bufSize)
 	go func() {
 		for {
```
Comment (robert-milan) [inline clock/clock.go]: Or we could just remove the default case, forcing this goroutine to block until it can send.

File: aggregator/aggregator.go (hunk 3 above)
Comment (Dieterbe) [reply aggregator/aggregator.go]: We've gone ahead with this PR because we need it.
These kinds of cleanups are nice to haves but we shouldn't spend too much time on it.
